graphistry/database_clients/SpannerGraph.py

```diff
@@ -0,0 +1,287 @@
+import os
+import pandas as pd
+import json
+import time
+import logging
+from typing import Any
+from google.cloud.spanner_dbapi.connection import connect
+from google.cloud.spanner_v1.data_types import JsonObject
+from google.cloud.spanner_v1 import KeySet
+
+logging.basicConfig(level=logging.INFO)
+
+class QueryResult:
+    """
+    Encapsulates the results of a query, including metadata.
+
+    Attributes:
+        data (list): The raw query results.
+        execution_time (float): The time taken to execute the query.
+        record_count (int): The number of records returned.
+    """
+
+    def __init__(self, data: list, execution_time: float):
+        self.data = data
+        self.execution_time = execution_time
+        self.record_count = len(data)
+
+    def summary(self) -> dict:
+        """
+        Provides a summary of the query execution.
+
+        Returns:
+            dict: A summary of the query results.
+        """
+        return {
+            "execution_time": self.execution_time,
+            "record_count": self.record_count
+        }
+
+class spannergraph:
+    """
+    A comprehensive interface for interacting with Google Spanner Graph databases. 
+
+    This class provides methods for connecting to the database, executing queries,
+    processing graph data into structured formats, and constructing visualizations
+    with the Graphistry library.
+
+    Attributes:
+        project_id (str): The Google Cloud project ID.
+        instance_id (str): The Spanner instance ID.
+        database_id (str): The Spanner database ID.
+        connection: The active connection to the Spanner database.
+        graphistry: The Graphistry parent object.
+
+    Key Methods:
+        __connect: Establish a connection to the Spanner database.
+        execute_query: Run a GQL query against the database.
+        parse_spanner_json: Convert raw Spanner JSON results into structured data.
+        get_nodes_df: Extract and structure graph nodes into a pandas DataFrame.
+        get_edges_df: Extract and structure graph edges into a pandas DataFrame.
+        gql_to_graph: Combine the above methods to generate a Graphistry visualization.
+        get_schema: Retrieve the schema of the database.
+        validate_data: Validate input data for queries or updates.
+        dump_config: Returns the configuration of the spannergraph instance.
+    """
```

Comment:
see the docs conventions in more modern code in pygraphistry, this is important for sphinx/readthedocs . likewise, for llm use, important to have examples.